[Not for merge] Sweeping ops for Dynamic work

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104733

Just using this op sweep to collect ops whose inputs are shape specialized with torch.compile.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ipiszy

anijain2305 committed:
Sweeping ops for Dynamic work

[ghstack-poisoned]

anijain2305 commented on test/dynamo/test_opinfo_dynamic.py:
```diff
@@ -0,0 +1,194 @@
+    correct = model(*ref_inputs, **ref_kwargs)
+    called = False
+
+    def detect_specialization(gm, name, *args):
+        for node in gm.graph.nodes:
+            example_value = node.meta.get("example_value", None)
+            if node.op == "placeholder" and isinstance(example_value, torch._subclasses.FakeTensor):
+                size = example_value.size()
+                for sz in size:
+                    if isinstance(sz, int) and sz not in (0, 1):
+                        # Found a specialization
+                        specialized_records.append([op, sz, gm])
+                        break
+            elif node.op == "placeholder" and isinstance(example_value, torch.SymInt):
+                si = node.meta["example_value"]
+                if len(torch.fx.experimental.symbolic_shapes.free_symbols(si)) == 0:
+                        # Found a specialization
+                        specialized_records.append([op, si, gm])
+                        break
```
@ezyang Does this look enough to catch specialization?

anijain2305 committed:
Update on "Sweeping ops for Dynamic work"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 committed:
Update on "[Not for merge] Sweeping ops for Dynamic work"


Just using this op sweep to collect ops whose inputs are shape specialized with torch.compile.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

ezyang commented on test/dynamo/test_opinfo_dynamic.py:
```diff
@@ -0,0 +1,208 @@
+            if node.op == "placeholder" and isinstance(example_value, torch._subclasses.FakeTensor):
+                size = example_value.size()
+                for sz in size:
+                    if isinstance(sz, int) and sz not in (0, 1):
```
This is insufficient. You may have a SymInt which actually has been specialized into an int. In fact, I would expect this to miss most specializations

ezyang commented:
If it's really only 80 ops that specialize right now, I think there's a decent chance that I'd argue we should add this as a permanent test.

anijain2305 committed:
Update on "[Not for merge] Sweeping ops for Dynamic work"


Just using this op sweep to collect ops whose inputs are shape specialized with torch.compile.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]
